### PR TITLE
DiagnosticVerifier: Implement defaulting behavior for lines in expected fix-it replacement range

### PR DESCRIPTION
**Pull request by AnthonyLatsis**
An audaciously late follow-up to https://github.com/apple/swift/pull/39572#discussion_r795089230.

* Default the start line to the line of the expected diagnostic.
* Default the end line to the start line.

Previously, the start and end lines of a replacement range were verified only when specified in the expected fix-it, e.g. `{{+1:5-10=bla}}`. This patch implements a sensible defaulting behavior for lines in an expected replacement range and thus enables unconditional rigorous full-range verification, which I believe is essential going forward. No noticeable difference is visible or expected performance-wise:
* We have fast paths for the common case of an intra-line source range (when the end location is on the same line).
* The LLVM source manager [exhaustively fills a line offset cache](https://github.com/apple/llvm-project/blob/43904c83c79ccd4eff61215cb5168af12ba46716/llvm/lib/Support/SourceMgr.cpp#L121) the first time a line number is requested in a given buffer. This already happens when the line of the expected diagnostic is computed.

Test changes are mostly cosmetic.

### CODE REVIEW

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test macOS

**Comment by AnthonyLatsis** [comment]
@rintaro ping

**Comment by AnthonyLatsis** [comment]
@rintaro ping

**Comment by rintaro** [comment]
Sorry for the delay. I will review this PR this evening.

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 1]
```diff
@@ -372,81 +358,85 @@ static Optional<LineColumnRange> parseExpectedFixItRange(
...
+    return ParsedLineAndColumn{FirstVal, Column};
   };
 
   LineColumnRange Range;
 
-  if (const auto lineAndCol = parseLineAndColumn()) {
-    std::tie(Range.StartLine, Range.StartCol) = lineAndCol.value();
+  if (const auto LineAndCol = parseLineAndColumn()) {
+    // The start line defaults to the line of the expected diagnostic.
+    Range.StartLine = LineAndCol->Line.getValueOr(DiagnosticLineNo);
```
Comment (rintaro) [inline lib/Frontend/DiagnosticVerifier.cpp:438]:
```suggestion
    Range.StartLine = LineAndCol->Line.value_or(DiagnosticLineNo);
```
`getValueOr` is deprecated.

File: lib/Frontend/DiagnosticVerifier.cpp [hunk 2]
```diff
@@ -459,8 +449,10 @@ static Optional<LineColumnRange> parseExpectedFixItRange(
     return None;
   }
 
-  if (const auto lineAndCol = parseLineAndColumn()) {
-    std::tie(Range.EndLine, Range.EndCol) = lineAndCol.value();
+  if (const auto LineAndCol = parseLineAndColumn()) {
+    // The end line defaults to the start line.
+    Range.EndLine = LineAndCol->Line.getValueOr(Range.StartLine);
```
Comment (rintaro) [inline lib/Frontend/DiagnosticVerifier.cpp:454]: ditto: `value_or`

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Review by rintaro — APPROVED** [review]
`DiagnosticVerifier` changes look good. Thank you!

File: lib/Frontend/DiagnosticVerifier.cpp (hunk 1 above)
Comment (AnthonyLatsis) [reply lib/Frontend/DiagnosticVerifier.cpp]: Thanks!

**Comment by AnthonyLatsis** [comment]
Note: Test changes is test case rearrangements (`test/Generics/associated_type_where_clause_hints.swift`; authored by me) and repositioning of verifications (the rest) for optimal offsets.

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test

**Pull request merged**